File: problems/scaler/stack/largest_rectangle_in_histogram.js
/**
 * Q. Largest Rectangle in Histogram

Given an array of integers A.

A represents a histogram i.e A[i] denotes the height of the ith histogram's bar. Width of each bar is 1.

Find the area of the largest rectangle formed by the histogram.

Problem Constraints
1 <= |A| <= 100000
1 <= A[i] <= 10000
*/
function soln(A) {
  function nextSmallerOnLeft(arr) {
    let stack = [];
    const n = arr.length;
    let rslt = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
      while (stack.length && arr[stack[stack.length - 1]] >= arr[i]) {
        stack.pop();
      }
      if (stack.length === 0) {
        rslt[i] = -1;
      } else {
        rslt[i] = stack[stack.length - 1];
      }
      stack.push(i);
    }

    return rslt;
  }

  function nextSmallerOnRight(arr) {
    let stack = [];
    const n = arr.length;
    let rslt = new Array(n).fill(0);
    for (let i = n - 1; i >= 0; i--) {
      while (stack.length && arr[stack[stack.length - 1]] >= arr[i]) {
        stack.pop();
      }
      if (stack.length === 0) {
        rslt[i] = n;
      } else {
        rslt[i] = stack[stack.length - 1];
      }
      stack.push(i);
    }

    return rslt;
  }

  let nextSmallerOnLeftArr = nextSmallerOnLeft(A);
  let nextSmallerOnRightArr = nextSmallerOnRight(A);

  let ans = 0;
  for (let i = 0; i < A.length; i++) {
    ans = Math.max(
      ans,
      (nextSmallerOnRightArr[i] - 1 - nextSmallerOnLeftArr[i]) * A[i]
    );
  }

  return ans;
}

console.log(soln([4, 3, 2, 1]));

File: src/main/java/com/dsa/problems/scaler/stack/largest_rectangle_in_histogram.java
package com.dsa.problems.scaler.stack;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Stack;

public class largest_rectangle_in_histogram {
  private static ArrayList<Integer> nextSmallerLeft(ArrayList<Integer> A) {
    Stack<Integer> st = new Stack<>();
    ArrayList<Integer> rslt = new ArrayList<>();
    for(int i = 0; i < A.size(); i++) rslt.add(0);

    for(int i = 0; i < A.size(); i++) {
      while(st.size() != 0 && A.get(st.peek()) >= A.get(i)) {
        st.pop();
      }
      if(st.size() == 0) {
        rslt.set(i, -1);
      } else {
        rslt.set(i, st.peek());
      }
      st.push(i);
    }

    return rslt;
  }

  private static ArrayList<Integer> nextSmallerRight(ArrayList<Integer> A) {
    Stack<Integer> st = new Stack<>();
    ArrayList<Integer> rslt = new ArrayList<>();
    for(int i = 0; i < A.size(); i++) rslt.add(0);

    for(int i = A.size() - 1; i >= 0; i--) {
      while(st.size() != 0 && A.get(st.peek()) >= A.get(i)) {
        st.pop();
      }
      if(st.size() == 0) {
        rslt.set(i, A.size());
      } else {
        rslt.set(i, st.peek());
      }
      st.push(i);
    }

    return rslt;
  }
 
  /**
 * Q. Largest Rectangle in Histogram

Given an array of integers A.

A represents a histogram i.e A[i] denotes the height of the ith histogram's bar. Width of each bar is 1.

Find the area of the largest rectangle formed by the histogram.

Problem Constraints
1 <= |A| <= 100000
1 <= A[i] <= 10000
*/
  private static int largestRectangleInHist(ArrayList<Integer> A) {
    ArrayList<Integer> nSL = nextSmallerLeft(A);
    ArrayList<Integer> nSR = nextSmallerRight(A);

    int ans = 0;
    for(int i = 0; i < A.size(); i++) {
      ans = Math.max(ans, (nSR.get(i) - 1 - nSL.get(i)) * A.get(i));
    }

    return ans;
  }

  public static void main(String[] args) {
    System.out.println(largestRectangleInHist(new ArrayList<>(Arrays.asList(4,3,2,1))));
  }
}
